### ISSUE TITLE
New check ChainedMethodCallWrap

### ISSUE THREAD

**Issue by nmancus1** [checkstyle/checkstyle#11087]
Noticed at https://github.com/checkstyle/checkstyle/pull/11081#discussion_r773266420 
 and https://github.com/checkstyle/checkstyle/pull/11083/files#r773320264 

It would be good to have an easy way to make sure that certain method calls are wrapped for readability. A good example from this repo is google truth assertions:

**How I would like check to work:**

input: 

```java
import static com.google.common.truth.Truth.assertWithMessage;
import org.junit.jupiter.api.Test;

public class TestClass {

    @Test
    void myTest() {
        final String actual = "some string";
        final String expected = "some string";
        assertWithMessage("some message") // ok
                .that(actual)
                .isEqualTo(expected);

        assertWithMessage("some message").that(actual).isEqualTo(expected); // violation
    }
}

```

config:

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="ChainedMethodCallWrap">
          <property name="methodNamePattern" value="^assert.*"/>
        </module>
    </module>
</module>


```

expected output: 

```bash

/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.2-all.jar -c config.xml TestClass.java
[ERROR] TestClass.java:14:42: '.' should be on new line. [ChainedMethodCallWrap]
[ERROR] TestClass.java:14:55: '.' should be on new line. [ChainedMethodCallWrap]


```

**Is your feature request related to a problem? Please describe.**
The problem is that no check/ combination of check and suppression easily allows enforcing of line wrapping chained method calls. 

**Describe the solution you'd like**
I would like a new check that accepts a method name pattern, so that common long chains of method calls (in assertions, builders, stream operations) can be validated for readability purposes.

**Comment by nmancus1** [comment]
My first thought was to use https://checkstyle.sourceforge.io/config_whitespace.html#OperatorWrap and some suppression for names that do not match, but this check does not accept the `DOT` token.

**Comment by Vyom-Yadav** [comment]
on it.

**Comment by pbludov** [comment]
Should  be there any violations in this code:

```java
Arrays.stream(cls.getMethods())                                         // static call, instance call 
  .filter(m -> m.getName().equals("get" + uppercaseFirstLetter(name)))  // lambda
  .findFirst()
  .map(Method::getReturnType)
  .orElseThrow(() -> new RuntimeException("Getter not found for field: " + name));
```
?

Another idea: we may need to specify the minimum number of call chains (maybe just one) after exceeding which there should be a break for readability:

```java
foo().bar(); // ok
foo().bar().baz(); // violation: chained method call should be wrapped
```

**Comment by Vyom-Yadav** [comment]
```
Arrays.stream(cls.getMethods())                                         // static call, instance call 
  .filter(m -> m.getName().equals("get" + uppercaseFirstLetter(name)))  // lambda
  .findFirst().....
```
@pbludov we cannot tell that a method is static or an instance method (AST based approach) so we cannot give an error based on that, lambda is the parameter so we can ignore it (the user may have different rules for methods inside lambda which will be applicable here).
>Should be there any violations in this code:

This can be customized, keep reading.
> Another idea: we may need to specify the minimum number of call chains (maybe just one) after exceeding which there should be a break for readability:

Agreed, along with the minimum number of chain calls, there should also be a property for the maximum number of chain calls, consider this example-
```
http.authorizeRequests()
                .antMatchers("/customer/delete*","/customer/showFormForAdd").hasRole("OWNER")
                .antMatchers("/customer/showFormForUpdate*").hasAnyRole("MANAGER", "OWNER")
                .antMatchers("/**").authenticated()
                    .and()
                .formLogin()
                .permitAll()
                    .and()
                .logout()
                .permitAll()
                    .and()
                .exceptionHandling()
                .accessDeniedPage("/access-denied");
```
It should be customizable for `antMatchers(..)` with minimum and the maximum number of chained calls on a single line as 2, also we can specify a property like `wrapFirstCall` which IntelliJ possesses. 
So to sum up we can have these properties-

- Regex for methods with the maximum and the minimum number of chained calls associated with these methods.
<pre>
[ pattern1,       pattern2, pattern3,pattern4 ]
      /\           /\             so on...
     /  \         /  \
    /    \       /    \
   /      \     /      \
 min      max min      max
</pre>
- A property called `wrapFirstCall`

The default value for min-max could be 1 or 2, as you suggest. Also, violation messages should be method name combined with dot as we want method along with dot on next or previous line
- `.methodA(..) should be on new line`
- `.methodA(..) should be on the previous line`

Previous line message is required as we can have a method call that was supposed to be on the previous line, only applicable if we go with the min-max rule.

**Comment by pbludov** [comment]
@nmancus1 @rnveach @romani @strkkk please share your thoughts

**Comment by nmancus1** [comment]
From https://github.com/checkstyle/checkstyle/issues/11087#issuecomment-1004812598: 
> Should be there any violations in this code..

It depends on user config. I had a very simple implementation in mind, which takes advantage of our treatment of the `DOT` and `METHOD_CALL` tokens (as operators/ parent tokens):

```
        |       |--EXPR -> EXPR [14:64]
        |       |   `--METHOD_CALL -> ( [14:64]
        |       |       |--DOT -> . [14:54]
        |       |       |   |--METHOD_CALL -> ( [14:46]
        |       |       |   |   |--DOT -> . [14:41]
        |       |       |   |   |   |--METHOD_CALL -> ( [14:25]
        |       |       |   |   |   |   |--IDENT -> assertWithMessage [14:8]

```

1. Check would use `METHOD_CALL` token.
2. Verify given regex in config against any method call with direct `IDENT` child (first method call in chain)
3. Check to make sure that each successive `DOT` ancestor is on a new line, report violations. We could get `IDENT` of method call that is a violation to show to user.

> Another idea: we may need to specify the minimum number of call chains... 

Yes, this is a good idea. Something like `maxSingleLineCalls` or something. Once we are over this number, all method calls should be wrapped.

> property for the maximum number of chain calls

I like the idea of a check like this, but I am not sure if it should be included in this check. To be consistent with our vision of atomicity in checks, I think this should be a new check itself. Maximum number of chained calls is not related to how such calls are line wrapped.

> we can specify a property like `wrapFirstCall`

This is a good property to have. `wrapFirstCall` makes formatting intention clear to user.

>  `pattern1, pattern2, pattern3, pattern4 `

I would encourage users to create separate modules for this case, instead of dealing with multiple patterns in a single check module/ instance.

**Comment by rnveach** [comment]
@pbludov I am biased as I was the one that suggested this check idea to @nmancus1 .

Usually builders are static calls since the builder is the one that will create the instance.

> `<property name="methodNamePattern" value="^assert.*"/>`

The only thing I wanted to point out is for this example we could allow the user to specify the class as the builder and not need for them to put in all the method name variations. `Truth` in this example is basically the builder and we can identify it's methods either from the static imports here or when its used as part of the method call (IE: `Truth.assertWithMessage`).

An example of a builder in my own project which are used to create test cases:
````
ClassUtil.get(requiredArguments).setDate(date).setTitle(title);
````

**Comment by Vyom-Yadav** [comment]
@nmancus1 @strkkk @rnveach I am ok with all suggestions and with using the class name instead of method name though we would lose some fine-grain control doing so as the user may not want to wrap every method in the builder class. I will start working on it as soon as it is approved.

**Comment by Vyom-Yadav** [comment]
@nmancus1 I am a little confused, are you pitching for 2 properties:

- `identifierPattern` - User inputs the name of the builder class `^ClassUtil.*`
- `methodName` - User inputs the method name **if they use star imports**, `^assertWithMessage`

**Comment by nmancus1** [comment]
> are you pitching for 2 properties

No, just one. Updated above comment to be more specific.

**Comment by Vyom-Yadav** [comment]
Oh ok if the user uses star imports then it is upon them to specify builder methods instead of class, seems good to me.

**Comment by nmancus1** [comment]
@romani first post updated with all details of new check proposal, please take a look.

**Comment by Vyom-Yadav** [comment]
@nmancus1 @rnveach  @pbludov  What about chained calls like-
```
((SomeOtherClazz) Clazz.get().setVal1(1)).someMethod1().someMethod2().build();
```
If we extract it into a variable-
```
SomeOtherClazz obj = (SomeOtherClazz) Clazz.get().setVal1(1);
obj.someMethod1().someMethod2().build(); 
```
These are two different method chains, all the properties should be separate for them?
What I mean is-
```
<module name="ChainedMethodCallWrap">
          <property name="identifierPattern" value="^.*"/>
          <property name="callsPerLine" value="1"/>
          <property name="maxSingleLineCalls" value="1"/>
</module>
```
```
((SomeOtherClazz) Clazz.get().setVal1(1)).someMethod1() // violation too many calls for single line chain (for the method chain inside parentheses).
    .someMethod2()
    .build();
```
for the same config-
```
((SomeOtherClazz) Clazz.get()
    .setVal1(1)).someMethod1().someMethod2().build(); // violation too many calls for single line chain (for method chain outside parentheses)
```
We need clear rules for chaining when done in parentheses (parentheses may be redundant or useful). Please share your opinion.

**Comment by nmancus1** [comment]
> We need clear rules for chaining when done in parentheses 

I think we have three options:

1. Anything under one `EXPR` parent is one method call (passing though all tokens , checking `DOT` and `METHOD_CALL` tokens until we reach `EXPR`). This would mean that your example 
```
((SomeOtherClazz) Clazz.get()
    .setVal1(1)).someMethod1().someMethod2().build();
```
would be one chain.
2. Consider any parent that is not a `DOT` or `METHOD_CALL` a "break" in the chain (the start of a new chain).
3. Parenthesis only is OK, but any other token will "break" the chain. In our AST, parenthesis are "parallel" to other tokens, so this would be easy to implement (we would still be just checking parents).


I am leaning towards (3). @Vyom-Yadav what do you think?

**Comment by Vyom-Yadav** [comment]
@nmancus1 I have implemented 1 in my local already, I didn't like it as it was breaking the rule of method chaining, hence this doubt was there.
> Parenthesis only is OK, but any other token will "break" the chain.

`INDEX_OP` will also be fine-
```
Clazz.get[0].foo();
```
```
|--EXPR -> EXPR [6:24]
|       |   `--METHOD_CALL -> ( 
|       |       |--DOT -> . 
|       |       |   |--INDEX_OP -> [ 
|       |       |   |   |--DOT -> . 
|       |       |   |   |   |--IDENT -> Clazz 
|       |       |   |   |   `--IDENT -> get 
|       |       |   |   |--EXPR -> EXPR
|       |       |   |   |   `--NUM_INT -> 0 
|       |       |   |   `--RBRACK -> ] 
|       |       |   `--IDENT -> foo 
|       |       |--ELIST -> ELIST 
|       |       `--RPAREN -> ) 
|       |--SEMI -> ; 
|       `--RCURLY -> } 
```
This other token will be `TYPECAST` in this case right? This will be a better solution, I will implement it.

```
((SomeOtherClazz) Clazz.get()
    .setVal1(1)).someMethod1().someMethod2().build();
```

```
|--EXPR -> EXPR 
|   `--METHOD_CALL -> ( 
|       |--DOT -> . 
|       |   |--METHOD_CALL -> ( 
|       |   |   |--DOT -> . 
|       |   |   |   |--METHOD_CALL -> ( 
|       |   |   |   |   |--DOT -> . 
|       |   |   |   |   |   |--LPAREN -> ( 
|       |   |   |   |   |   |--TYPECAST -> ( 
|       |   |   |   |   |   |   |--TYPE -> TYPE 
|       |   |   |   |   |   |   |   `--IDENT -> SomeOtherClazz 
|       |   |   |   |   |   |   |--RPAREN -> ) 
|       |   |   |   |   |   |   `--METHOD_CALL -> ( 
|       |   |   |   |   |   |       |--DOT -> . 
|       |   |   |   |   |   |       |   |--METHOD_CALL -> ( 
|       |   |   |   |   |   |       |   |   |--DOT -> . 
|       |   |   |   |   |   |       |   |   |   |--IDENT -> Clazz 
|       |   |   |   |   |   |       |   |   |   `--IDENT -> get 
|       |   |   |   |   |   |       |   |   |--ELIST -> ELIST 
|       |   |   |   |   |   |       |   |   `--RPAREN -> ) 
|       |   |   |   |   |   |       |   `--IDENT -> setVal1 
|       |   |   |   |   |   |       |--ELIST -> ELIST 
|       |   |   |   |   |   |       |   `--EXPR -> EXPR
|       |   |   |   |   |   |       |       `--NUM_INT -> 1 
|       |   |   |   |   |   |       `--RPAREN -> )
|       |   |   |   |   |   |--RPAREN -> ) 
|       |   |   |   |   |   `--IDENT -> someMethod1
|       |   |   |   |   |--ELIST -> ELIST
|       |   |   |   |   `--RPAREN -> ) 
|       |   |   |   `--IDENT -> someMethod2
|       |   |   |--ELIST -> ELIST 
|       |   |   `--RPAREN -> ) 
|       |   `--IDENT -> build 
|       |--ELIST -> ELIST 
|       `--RPAREN -> ) 
|--SEMI -> ; 
`--RCURLY -> } 
```

**Comment by nmancus1** [comment]
> INDEX_OP will also be fine-

Yes, good point.

I would consider the typecast to break the "chain" in this case.

I am approving new check. If implementation does not pass all maintainers, we can always move it to sevntu to let it mature. We can even do the same (move to senvtu) if we end up with issues/ controversy after release.

@Vyom-Yadav please proceed, just be aware that this may end up living in sevntu for some time if all maintainers do not approve of check.

**Comment by Vyom-Yadav** [comment]
@nmancus1 The more I read about it the more ambiguity it creates. Most of the definitions I found stated that method chaining is a technique in which multiple method calls are made in one **statement**. Technically -
```
((SomeOtherClazz) Clazz.get()
    .setVal1(1)).someMethod1().someMethod2().build();
```
is one statement. Many users might not like this behavior but some of them may like it, can we make a property for such type of behavior that satisfies both types of users?

**Comment by nmancus1** [comment]
> The more I read about it the more ambiguity it creates.

Yes, I have been thinking more about comments above, and looking through current PR. Let’s do (1) from https://github.com/checkstyle/checkstyle/issues/11087#issuecomment-1039781408. One statement, or one expression, is one method call chain. This gives us a clear “stopping” point when ascending through AST. We will let users dictate in the future whether we need to consider some other option/property.

**Comment by Vyom-Yadav** [comment]
@nmancus1 Some more cases in which I need some input-
```
((asd() + someVar).sdsf() + anotherVar).asd().asd();
```
Now it is hard to visualize this as a chain in general, this is just an example (not compilable) but other cases are possible in which expression are evaluated which result in an object on which method is called.
Should this also be considered as a single chain?

**Comment by nmancus1** [comment]
> not compilable

Let's find concrete example and go from there. I think that this would still be "one chain", since it should be under one `EXPR` node.

**Comment by nmancus1** [comment]
>  how do I possibly name this method chain (1)?

Can we just skip any method call chain where the "innermost" (AST-wise) method call does not have a `DOT` first child with two `IDENT`s as children?

**Comment by Vyom-Yadav** [comment]
@nmancus1 I was able to get some workaround logic-
```
public String test() {
    int pc = 0;
    return ((100000+pc)+"abc".substring(0,2)).substring(1); // ------- (1)
}
```
`"abc".substring(0,2)` is a different chain, `((100000+pc)+"abc".substring(0,2)).substring(1)` is a different chain, it is effectively, `someObj.substring(1)` as user will specify this chain as `.*\.substring`.

Logic development is complete (as of now)  some more test cases will be added, few regression reports with a unique pattern (like - `assert.*`) is pending, you can perform an initial review if you wish.

Also if you have any other configurations (`identifierPattern` for any other project) for any projects that use method chaining extensively, feel free to mention them.

**Comment by lisp-ceo** [comment]
Could this get merged? We're looking to adopt it.

**Comment by rnveach** [comment]
Anyone is free to finish the PR that was closed ( https://github.com/checkstyle/checkstyle/pull/11231 ) and try to get this accepted and merged.

**Comment by romani** [comment]
Last comment was about to move it to https://github.com/sevntu-checkstyle/sevntu.checkstyle ,  we just need to find someone to copy code to other repo.

**Comment by RicardoMonteiroSimoes** [comment]
Looks like that status of this is that nobody is working on, so Checkstyle has no support as off right now for builder pattern formatting?

**Comment by romani** [comment]
We are waiting for help to move referenced implementation to https://github.com/sevntu-checkstyle/sevntu.checkstyle , to let us release it and get feedback on usage of it, as there are nuances.

**Comment by walker99011** [comment]
@romani is moving referenced implementation to https://github.com/sevntu-checkstyle/sevntu.checkstyle still actual? If code is moved we need to wait for release then? Looks like sevntu has not been released for some time and there is no compatibility with Checkstyle 12.3.0 (actual for 18.12.25).

**Comment by romani** [comment]
> is moving referenced implementation to https://github.com/sevntu-checkstyle/sevntu.checkstyle still actual? 

yes

>If code is moved we need to wait for release then?

I will release it right after code is merged.

>Looks like sevntu has not been released for some time 

there is no functional fixes merged, so no releases.
